students/HarveyHerela/lesson04/mailroom2.py

```diff
@@ -0,0 +1,130 @@
+          {"firstname": "Skye", "lastname": "Rubble", "donations": [100, 100, 100]}]
+
+def get_name(name_type):    
+    gettingInput = True
+
+    while gettingInput:
+        
+        print("Type 'list' for a list of names in the database, '~' to go back")
+        if name_type == "first":
+            response = input("First name: ")
+        elif name_type == "last":
+            response = input("Last name: ")
+
+        if response.lower() == "list":
+            for d in donors:
+                print("{firstname} {lastname}".format(**d))
+        else:
+            gettingInput = False
+    return response
+
+def send_thank_you():
+    """Prompts for a donor name and donation amount, then prints a 'thank you' message to the screen."""
+    firstname= get_name("first")
+
+    if firstname == "~":
+        return
+
+    lastname = get_name("last")
+
+    if lastname == "~":
+        return
+
+    print("Donation amount, or '~' to go to main menu:")
+    amount = input(":")
+    if amount == '~':
+        return
+    amount = float(amount)
+    # If it's an existing donor, add it to their donation list.
+    # If it's a new donor, create a new entry.
+    for d in donors:
+        if d["firstname"] == firstname and d["lastname"] == lastname:
+            d["donations"].append(amount)
+            break
+    else:
+        donors.append({"firstname": firstname, "lastname": lastname, "donations": [amount]})
+        
+    thank_you_msg = "Thank you {firstname} {lastname} for your generous donation of ${amount:.2f}".format(
+        firstname=firstname, lastname=lastname, amount=amount)
+    print(thank_you_msg)
+
+def get_total(donor_data):
+    return donor_data[1]
+
+def create_report():
+    """Prints a list of all donors, sorted by total amount given."""
+    report = str()
+    report += "Donor Name    | Total Given   | Num Gifts | Average Gift"
+    report += "\n"
+    report += "--------------------------------------------------------" 
+    report += "\n"
+    # Create a list with the data to be printed
+    data_list = list()
+    for donor_data in donors:
+        fullname = "{firstname} {lastname}".format(**donor_data)
+        gift_data = [fullname, 0, 0, 0]
+        for donation in donor_data["donations"]:
+            gift_data[1] += donation
+            gift_data[2] += 1
+        gift_data[3] = gift_data[1] / gift_data[2]
+        data_list.append(gift_data)
+    
+    # Sort the list by total amount given
+    data_list.sort(key=get_total, reverse=True)
+    
+    # Define the formatter
+    line_fmtr = "{name:<14}  ${total:>13.2f}   {num:>9d}  ${avg:>12.2f}\n".format
+    for d in data_list:
+        report += line_fmtr(name=d[0], total=d[1], num=d[2], avg=d[3])
+    
+    print("\n", report)
+
+def send_letters_everyone():
+    """Creates a letter for everyone in the database, and writes them to file."""
+    num = 0
+    subdir = str(datetime.date.today())
```

Comment:
You could use `datetime.datetime.now()` here so that a new directory would be created every time the function is called (in this version the user is limited to one chance per day to send letters). Or better yet, use a timestamp on the letters themselves and just keep adding them to the existing folder.

It's still a great idea to use a `try/except` block here, but instead of trying to guess what the specific error might be, you could have a more generic error message for your user and log the specific error for reference. The simplest way to 'log' an exception is just to use a print statement by calling `print(oserr)` inside your try block, or you could pass `osser` to a format string for a more user friendly error message.